sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java

```diff
@@ -138,16 +147,115 @@ public void onFailure(Throwable e) {
-            .includeGlobalState(false)
-            .includeAliases(true);
         FutureActionListener<RestoreSnapshotResponse, Long> listener = new FutureActionListener<>(Functions.constant(1L));
-        transportActionProvider.transportRestoreSnapshotAction().execute(request, listener);
+        resolveIndexNames(analysis.restoreTables(), ignoreUnavailable, transportActionProvider.transportGetSnapshotsAction(), analysis.repositoryName())
+            .whenComplete((List<String> indexNames, Throwable t) -> {
+                if (t == null) {
+                    RestoreSnapshotRequest request = new RestoreSnapshotRequest(analysis.repositoryName(), analysis.snapshotName())
+                        .indices(indexNames)
+                        .indicesOptions(indicesOptions)
+                        .settings(analysis.settings())
+                        .waitForCompletion(waitForCompletion)
+                        .includeGlobalState(false)
+                        .includeAliases(true);
+                    transportActionProvider.transportRestoreSnapshotAction().execute(request, listener);
+                } else {
+                    listener.onFailure(t);
+                }
+            });
         return listener;
     }
+
+    @VisibleForTesting
+    static CompletableFuture<List<String>> resolveIndexNames(List<RestoreSnapshotAnalyzedStatement.RestoreTableInfo> restoreTables, boolean ignoreUnavailable,
+                                                             TransportGetSnapshotsAction getSnapshotsAction, String repositoryName) {
+        RestoreIndices resolvedIndices = new RestoreIndices();
+        List<RestoreSnapshotAnalyzedStatement.RestoreTableInfo> toResolveFromSnapshot = new ArrayList<>();
+        for (RestoreSnapshotAnalyzedStatement.RestoreTableInfo table : restoreTables) {
+            if (table.hasPartitionInfo()) {
+                resolvedIndices.add(table.partitionName().asIndexName());
+            } else if (ignoreUnavailable) {
+                // If ignoreUnavailable is true, it's cheaper to simply return indexName and the partitioned wildcard instead
+                // checking if it's a partitioned table or not
+                resolvedIndices.add(table.tableIdent().indexName());
+                resolvedIndices.add(PartitionName.templateName(table.tableIdent().schema(), table.tableIdent().name()) + "*");
+            } else {
+                // index name needs to be resolved from snapshot
+                toResolveFromSnapshot.add(table);
+            }
+        }
+        if (toResolveFromSnapshot.isEmpty()) {
+            return CompletableFuture.completedFuture(resolvedIndices.toList());
+        }
+        final CompletableFuture<List<String>> f = new CompletableFuture<>();
+        getSnapshotsAction.execute(
+            new GetSnapshotsRequest(repositoryName),
+            new ResolveFromSnapshotActionListener(f, toResolveFromSnapshot, resolvedIndices)
```

Comment:
can't use that, in it's current implementation it doesn't handle exceptions thrown within the transform function.